sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuRegExpReplaceMeta.scala

```diff
@@ -81,19 +124,53 @@ class GpuRegExpReplaceMeta(
     // ignore the pos expression which must be a literal 1 after tagging check
     require(childExprs.length == 4,
       s"Unexpected child count for RegExpReplace: ${childExprs.length}")
-    if (canUseGpuStringReplace) {
-      GpuStringReplace(lhs, regexp, rep)
-    } else {
-      (javaPattern, cudfPattern, replacement) match {
-        case (Some(javaPattern), Some(cudfPattern), Some(cudfReplacement)) =>
-          if (containsBackref) {
-            GpuRegExpReplaceWithBackref(lhs, cudfPattern, cudfReplacement)
-          } else {
-            GpuRegExpReplace(lhs, regexp, rep, javaPattern, cudfPattern, cudfReplacement)
-          }
-        case _ =>
-          throw new IllegalStateException("Expression has not been tagged correctly")
-      }
+
+    action match {
+      case Some(ActionStringReplace) =>
+        GpuStringReplace(lhs, regexp, rep)
+      case Some(ActionStringReplaceMulti(patterns, javaPattern, cudfPattern, replacement))
+          if patterns.length == 2 =>
+        // We can't just replace the two strings in turn because it can lead to incorrect
+        // results, so we first replace string1 with a placeholder that is guaranteed to not
+        // appear in the string, then we replace string2 and finally replace the placeholder.
+        //
+        // Example:
+        // - input string "cabd"
+        // - pattern "ab|cd"
+        // - replacement string ""
+        //
+        // step 1: replace "ab" with "^", resulting in "c^d"
+        // step 2: replace "cd" with "^", resulting in "c^d"
+        // step 3: replace "^" with "", resulting in "cd"
+        //
+        // We also have to check if the input data contains the placeholder and fall back to
+        // regexp_replace in that case. We artificially declare that the expression has
+        // side-effects to trigger the gather/scatter logic so that we only invoke
+        // regexp_replace when needed
+        val placeholder = GpuLiteral("^")
+
+        // first replace string 1 with a special placeholder
+        val repl1 = GpuStringReplace(lhs, GpuLiteral(patterns.head), placeholder)
+        // now replace string 2
+        val repl2 = GpuStringReplace(repl1, GpuLiteral(patterns(1)), rep)
+        // finally replace the placeholder
+        // declare that this has side-effects to force the scatter-gather algorithm
+        val stringReplace = GpuStringReplace(repl2, placeholder, rep, hasSideEffects = true)
+
+        val regexpReplace = GpuRegExpReplace(
+          lhs, regexp, rep, javaPattern, cudfPattern, replacement)
+
+        val containsPlaceholder = GpuRLike(lhs, GpuLiteral("\\^"), "\\^")
+
+        GpuIf(containsPlaceholder, regexpReplace, stringReplace)
```

Comment:
I really don't like using a GpuIf here and marking part of it as havning a side effect as an attempt to make this perform better. But I don't see any performance numbers yet in the PR.  I think this is a premature optimization step. Could we get some numbers looking at how long the regexp_replace takes today with various lengths of input data. And compare that to the very simple use case of no overlap in the two parts of the or so all we do is a string replace twice.

I would really like to know how fast that is compared to what we have today, and if there is a heuristic we can use to decide how we want to use this.  I really think that we have to build a lot of this into `GpuRegexpReplace` itself because my gut is telling me that the normal regexp replace is going to be faster for small inputs strings, and much slower for large ones. And we don't know that until we actually see the data.

Along with that the CUDF team mentioned offline that we might be able to use this version of string replace

https://github.com/rapidsai/cudf/blob/d077c9b2a3764818fb690995f2c2867a2c024811/cpp/include/cudf/strings/replace.hpp#L150-L154

to speed it up even more because then we would only be producing a single output string column instead of producing an intermediate string column followed by a final string column. I don't know if this has been optimized yet for large strings. That might be a big downside to using this API.  But if it is optimized for the large strings case, then we could also do some testing to see what the performance impact is for more than static strings with or in between them. i.e. "ab|cd|ef".

I would very much like to see us wait on trying to tackle the overlapping pattern use case until we have some performance numbers to compare it with so we know if it is worth it.

Reply:
I just finished benchmarking with some sample data I was provided with, and the approach in this PR does not help at all. My test takes around 34 minutes, and performance was the same (within 1%) with the three string replace operations (without the GpuIf) compared to the original regexp_replace.

Reply:
Is your benchmark data skew (i.e., has both short+long strings)? That should be the case of interest for this PR.

Reply:
In addition, did you consult with David Wendt to check if `strings::replace` has been optimized for long strings? We had performance issue with `split_record` with long strings until we requested to optimize it.

Reply:
I looked at the code for the multi-replace string op that we would want to use and it has not been optimized for long strings.